app/controllers/api/storage.php

```diff
@@ -927,41 +1044,95 @@
 
         $path = $file->getAttribute('path', '');
 
-        if (!\file_exists($path)) {
-            throw new Exception('File not found in '.$path, 404);
+        $device = Storage::getDevice('files');
+        
+        if (!$device->exists($path)) {
+            throw new Exception('File not found in ' . $path, 404);
         }
 
-        $device = Storage::getDevice('files');
+        $usage
+            ->setParam('storage.files.read', 1)
+            ->setParam('bucketId', $bucketId)
+        ;
 
-        $source = $device->read($path);
+        $response
+            ->setContentType($file->getAttribute('mimeType'))
+            ->addHeader('Expires', \date('D, d M Y H:i:s', \time() + (60 * 60 * 24 * 45)) . ' GMT') // 45 days cache
+            ->addHeader('X-Peak', \memory_get_peak_usage())
+            ->addHeader('Content-Disposition', 'attachment; filename="' . $file->getAttribute('name', '') . '"')
+        ;
+
+        $size = $file->getAttribute('sizeOriginal', 0);
+
+        $rangeHeader = $request->getHeader('range');
+        if(!empty($rangeHeader)) {  
+            list($unit, $range) = explode('=', $rangeHeader);
+            if($unit == 'bytes' && !empty($range)) {
+                list($rangeStart, $rangeEnd) = explode('-', $range);
+                if(strlen($rangeStart) == 0 || strstr($range, '-') === false) {
+                    throw new Exception('Invalid range', 416);
+                }
+                $rangeStart = (int) $rangeStart;
+                if(strlen($rangeEnd) == 0) {
+                    $rangeEnd =  min(($rangeStart + 2000000-1), ($size - 1));
+                } else {
+                    $rangeEnd = (int) $rangeEnd;
+                }
+                if(($rangeStart >= $rangeEnd) || $rangeEnd >= $size) {
+                    throw new Exception('Invalid range', 416);
+                }
+                
+                $response
+                    ->addHeader('Accept-Ranges', 'bytes')
+                    ->addHeader('Content-Range', 'bytes ' . $rangeStart . '-' . $rangeEnd . '/' . $size)
+                    ->addHeader('Content-Length', $rangeEnd - $rangeStart + 1)
+                    ->setStatusCode(Response::STATUS_CODE_PARTIALCONTENT);
+            } else {
+                throw new Exception('Invalid range', 416);
+            }
+        }
+
+        $source = '';
         if (!empty($file->getAttribute('openSSLCipher'))) { // Decrypt
+            $source = $device->read($path);
             $source = OpenSSL::decrypt(
                 $source,
```

Comment:
```suggestion
            $source = OpenSSL::decrypt(
                $device->read($path),
```

We can probably skip this assignment 👍🏻

Reply:
We can't as it's required in next step, when encryption is disabled, next step will fail.

Reply:
Ah okay 👍🏻